aten/src/ATen/native/sparse/SoftMax.cpp

```diff
@@ -0,0 +1,467 @@
+#include <ATen/ATen.h>
+#include <ATen/Config.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/SparseTensorUtils.h>
+#include <ATen/Parallel.h>
+#include <ATen/NamedTensorUtils.h>
+#include <map>
+
+namespace at {
+namespace native {
+namespace {
+
+template <typename iscalar_t>
+iscalar_t get_nvalues(const std::vector<iscalar_t>& sizes, const int64_t sparse_dim) {
+  /* Return the number of elements of sparse values entries.
+   */
+  iscalar_t nvalues = 1;
+  for (auto it = sizes.begin() + sparse_dim; it != sizes.end(); ++it) {
+    nvalues *= *it;
+  }
+  return nvalues;
+}
+
+template <typename iscalar_t>
+std::vector<iscalar_t> get_offsets(const Tensor& indices, const std::vector<iscalar_t>& sizes, const int64_t dim) {
+  /*
+    Given the indices of a sparse tensor return a vector of offsets
+    for the entries in the corresponding dense tensor:
+
+      If
+        offsets = get_offsets(A._indices(), A.sizes(), -1)
+        data = A.to_dense().resize((nnz,))
+      then
+        data[offsets[n]] == A._values()[n]
+
+    `indices` must be a contiguous 2-d tensor with iscalar_t entries.
+    `sizes` must be a vector with at least ndim entries.
+
+    `dim` is an integer. When >= 0 and < ndim, all entries in the
+    given dimension will be mapped to the first entry. Otherwise, the
+    value is ignored.
+
+    The items in the returned vector are in increasing order.
+  */
+  auto ndim = indices.size(0);
+  auto nnz = indices.size(1);
+  std::vector<iscalar_t> offsets(nnz, 0);
+  std::vector<iscalar_t> strides(ndim, 1);
+  iscalar_t* indices_data_base = indices.data_ptr<iscalar_t>();
+
+  if (ndim > 1) {
+    for (int64_t i=ndim - 2; i >= 0; i--) {
+      strides[i] = strides[i + 1] * sizes[i + 1];
+    }
+  }
+
+  for (int64_t j=0; j < ndim; j++) {
+    if (j != dim) {
+      for (int64_t i=0; i < nnz; i++) {
+        offsets[i] += strides[j] * indices_data_base[j * nnz + i];
+      }
+    }
+  }
+
+  return offsets;
+}
+
+template <typename iscalar_t>
+std::vector<iscalar_t> get_dense_offsets(iscalar_t &mx, const std::vector<iscalar_t>& offsets) {
+  /*
+    Return the dense set of offsets:
+
+      If
+        pool = get_dense_offsets(mx, offsets)
+      then
+        pool.size() == offsets.size()
+        pool[i] == pool[j] iff offsets[i] == offsets[j]
+        set(pool) == set(range(mx))
+  */
+  auto n = offsets.size();
+  std::vector<iscalar_t> pool(n, 0);
+  std::map<iscalar_t, iscalar_t> i2p;
+  for (int64_t i=0; i < n; i++) {
+    auto c = offsets[i];
+    auto it = i2p.find(c);
+    iscalar_t p = i2p.size();
+    if (it == i2p.end()) {
+      i2p.emplace(std::make_pair(c, p));
+    } else {
+      p = it->second;
+    }
+    pool[i] = p;
+  }
+  mx = i2p.size();
+  return pool;
+}
+
+template <typename scalar_t, typename iscalar_t, bool LogSoftMax>
```

Comment:
nit: I think that index is always of type `int64_t`.

Reply:
renamed `iscalar_t` to `index_t`.